# import Library

In [1]:
import torch
from torch import nn
import torch.nn.functional as F
from torch.utils.data import Dataset
import torch.optim as optim
import torchvision
import torchvision.transforms as transforms
from torch.utils.tensorboard import SummaryWriter

from PIL import Image
from collections import OrderedDict #for multiGPU
from tqdm import tqdm

import os
import random
import time

# GPU

In [2]:
# device = torch.device("mps") if torch.backends.mps.is_available() else "cpu"
#for colab
device = 'cuda' if torch.cuda.is_available() else 'cpu'
print("device is ", device)

device is  cuda


# Load Data

In [3]:
print("loading dataset")

data_transforms = {
    'train' : transforms.Compose([ #horizontal flip, random crop, normalization           
            #transforms.ToPILImage(),
            # transforms.Resize((256,256)),
            transforms.RandomCrop(32, padding = 4), #augmentation, input for ResNet
            transforms.RandomHorizontalFlip(), #augmentation
            transforms.ToTensor(),
            transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
            #transforms.Normalize((0.4914, 0.4822, 0.4465), (1, 1, 1)), #with the per-pixel mean subtracted

        ]) , 
        'test': transforms.Compose([
            transforms.ToTensor(), #single view of the original 32x32 image
            transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225)),
        ])
    }
#access data_transforms['train']

#dataloader (batch)
train_set = torchvision.datasets.CIFAR10(
    root='./data', train=True, download=True, transform = data_transforms['train']
)
trainloader = torch.utils.data.DataLoader(
    train_set, batch_size = 128, shuffle = True, num_workers=2
)

test_set = torchvision.datasets.CIFAR10(
    root='./data', train=False, download=True, transform=data_transforms['test'])
testloader = torch.utils.data.DataLoader(
    test_set, batch_size=100, shuffle=False, num_workers=2)

#classes
classes = ('plane', 'car', 'bird', 'cat', 'deer','dog', 'frog', 'horse', 'ship', 'truck')


loading dataset


100%|██████████| 170498071/170498071 [00:03<00:00, 49024702.78it/s]


Extracting ./data/cifar-10-python.tar.gz to ./data
Files already downloaded and verified


# Model Architecture

In [ ]:
class optionA(nn.Module):
  def __init__(self, func):
    super(optionA, self).__init__()
    self.func = func
  
  def forward(self, x):
    return self.func(x)

In [4]:
class Fx2_16(nn.Module):
  def __init__(self):
    super().__init__()
    self.conv1 = nn.Sequential(
        nn.Conv2d(16, 16, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(16)
    )
    self.conv2 = nn.Sequential(
        nn.Conv2d(16, 16, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(16)
    )
    self.residual = nn.Sequential()
  def forward(self, x):
    out = F.relu(self.conv1(x))
    out = self.conv2(out)
    out +=self.residual(x)
    out = F.relu(out)

    return out

In [5]:
class Fx2_32(nn.Module):
  def __init__(self):
    super().__init__()
    self.conv1 = nn.Sequential(
        nn.Conv2d(32, 32, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(32)
    )
    self.conv2 = nn.Sequential(
        nn.Conv2d(32, 32, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(32),
    )
    self.residual = nn.Sequential()
  
  def forward(self, x):
    out = F.relu(self.conv1(x))
    out = self.conv2(out)
    out += self.residual(x)
    out = F.relu(out)
    return out


In [6]:
class Fx2_64(nn.Module):
  def __init__(self):
    super().__init__()

    self.conv1 = nn.Sequential(
         nn.Conv2d(64, 64, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(64)
    )
    self.conv2 = nn.Sequential(
        nn.Conv2d(64, 64, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(64)
    )
    self.residual = nn.Sequential()

  def forward(self, x):
    out = F.relu(self.conv1(x))
    out = self.conv2(out)
    out += self.residual(x)
    out = F.relu(out)
    return out

In [7]:
#classes for Resnet56 Model
class conv16 (nn.Module): #Blue
  def __init__(self):
    super().__init__()
    self.Fx1 = Fx2_16()
    self.Fx2 = Fx2_16()
    self.Fx3 = Fx2_16()
    self.Fx4 = Fx2_16()
    self.Fx5 = Fx2_16()
    self.Fx6 = Fx2_16()
    self.Fx7 = Fx2_16()
    self.Fx8 = Fx2_16()
    self.Fx9 = Fx2_16()
    
  def forward(self,x):
    x1 = self.Fx1(x)
    x2 = self.Fx2(x1)
    x3 = self.Fx3(x2) 
    x4 = self.Fx4(x3) 
    x5 = self.Fx5(x4) 
    x6 = self.Fx6(x5) 
    x7 = self.Fx7(x6) 
    x8 = self.Fx8(x7) 
    x9 = self.Fx9(x8)
    #x = self.relu(x)
    return x9

#classes for Resnet56 Model
class conv32 (nn.Module): #Purple
  def __init__(self):
    super().__init__()
    self.Fx1_1 = nn.Sequential(
        #torch.nn.Conv2d(in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1, groups=1, bias=True, padding_mode='zeros', device=None, dtype=None)
        nn.Conv2d(16, 32, kernel_size = 3, stride = 2, padding = 1, bias = False),
        nn.BatchNorm2d(32))
    self.Fx1_2 = nn.Sequential(
        nn.Conv2d(32, 32, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(32)
    )

    ## OptionB
    self.residual_mapping = nn.Sequential(
        nn.Conv2d(16,32,kernel_size =1, stride = 2, bias= False),
        nn.BatchNorm2d(32)
    )

    # ## OptionA
    #self.residual_mapping = optionA(lambda x: F.pad(x[:, :, ::2, ::2], (0, 0, 0, 0, 32//4, 32//4), "constant", 0))

    self.Fx2 = Fx2_32()
    self.Fx3 = Fx2_32()
    self.Fx4 = Fx2_32()
    self.Fx5 = Fx2_32()
    self.Fx6 = Fx2_32()
    self.Fx7 = Fx2_32()
    self.Fx8 = Fx2_32()
    self.Fx9 = Fx2_32()
    # self.relu = nn.ReLU()
    
  def forward(self,x):
    out = F.relu(self.Fx1_1(x))
    out = self.Fx1_2(out) 
    out += self.residual_mapping(x)
    x1 = F.relu(out)
    
    x2 = self.Fx2(x1) 
    x3 = self.Fx3(x2) 
    x4 = self.Fx4(x3) 
    x5 = self.Fx5(x4) 
    x6 = self.Fx6(x5) 
    x7 = self.Fx7(x6) 
    x8 = self.Fx8(x7) 
    x9 = self.Fx9(x8) 
    # x2 = self.Fx2(x1) + x1
    # x3 = self.Fx3(x2) + x2
    #x = self.relu(x)
    return x9

#classes for Resnet56 Model
class conv64 (nn.Module): #Green
  def __init__(self):
    super().__init__()
    self.Fx1_1 = nn.Sequential(
        #torch.nn.Conv2d(in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1, groups=1, bias=True, padding_mode='zeros', device=None, dtype=None)
        nn.Conv2d(32, 64, kernel_size = 3, stride = 2, padding = 1, bias = False),
        nn.BatchNorm2d(64)
    )
    self.Fx1_2 = nn.Sequential(
        nn.Conv2d(64, 64, kernel_size = 3, stride = 1, padding = 1, bias = False),
        nn.BatchNorm2d(64)
    )

    ## OptionB
    self.residual_mapping = nn.Sequential(
        nn.Conv2d(32,64,kernel_size =1, stride = 2, bias= False),
        nn.BatchNorm2d(64)
    )

    ## OptionA
    #self.residual_mapping = optionA(lambda x: F.pad(x[:, :, ::2, ::2], (0, 0, 0, 0, 64//4, 64//4), "constant", 0))
    
    self.Fx2 = Fx2_64()
    self.Fx3 = Fx2_64()
    self.Fx4 = Fx2_64()
    self.Fx5 = Fx2_64()
    self.Fx6 = Fx2_64()
    self.Fx7 = Fx2_64()
    self.Fx8 = Fx2_64()
    self.Fx9 = Fx2_64()

    # self.relu = nn.ReLU()
    
  def forward(self,x):
    out = F.relu(self.Fx1_1(x))
    out = self.Fx1_2(out)
    out += self.residual_mapping(x)
    x1 = F.relu(out)
    x2 = self.Fx2(x1) 
    x3 = self.Fx3(x2) 
    x4 = self.Fx4(x3) 
    x5 = self.Fx5(x4)
    x6 = self.Fx6(x5) 
    x7 = self.Fx7(x6) 
    x8 = self.Fx8(x7) 
    x9 = self.Fx9(x8) 

    return x9
   


In [8]:
class MyModel(nn.Module): #Resnet56 architecture implementation
    def __init__(self, conv2 = conv16, conv3 = conv32,
            conv4 = conv64, num_classes = 10, init_weights = True): #num_classes
        super().__init__()
        self.conv1 = nn.Sequential(
            nn.Conv2d(3, 16, kernel_size = 3, stride = 1, padding=1, bias = False),
            nn.BatchNorm2d(16)
        )
        self.conv2 = conv2()
        self.conv3 = conv3()
        self.conv4 = conv4()
        #self.avg_pool = nn.AdaptiveAvgPool2d((1,1))
        self.fclayer = nn.Linear(64, num_classes)

        if init_weights:
            self.initialize_weights()
    
    # define weight initialization function
    def initialize_weights(self):
        for m in self.modules():
            if isinstance(m, nn.Conv2d):
                nn.init.kaiming_normal_(m.weight, mode='fan_out', nonlinearity='relu')
                if m.bias is not None:
                    nn.init.constant_(m.bias, 0)
            elif isinstance(m, nn.BatchNorm2d):
                nn.init.constant_(m.weight, 1)
                nn.init.constant_(m.bias, 0)
            elif isinstance(m, nn.Linear):
                nn.init.normal_(m.weight, 0, 0.01)
                nn.init.constant_(m.bias, 0) 
        
    def forward(self, x):
        x1 = F.relu(self.conv1(x))
        x2 = self.conv2(x1)
        x3 = self.conv3(x2)
        x4 = self.conv4(x3)
        x = F.avg_pool2d(x4, x4.size()[3])
        #x = self.avg_pool(x4)
        x = x.view(x.size(0), -1)
        output = self.fclayer(x)
        return output

In [ ]:
resnet56 = MyModel()

In [ ]:
from torchsummary import summary
summary(resnet56, (3,32,32))

----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1           [-1, 16, 32, 32]             432
       BatchNorm2d-2           [-1, 16, 32, 32]              32
            Conv2d-3           [-1, 16, 32, 32]           2,304
       BatchNorm2d-4           [-1, 16, 32, 32]              32
            Conv2d-5           [-1, 16, 32, 32]           2,304
       BatchNorm2d-6           [-1, 16, 32, 32]              32
            Fx2_16-7           [-1, 16, 32, 32]               0
            Conv2d-8           [-1, 16, 32, 32]           2,304
       BatchNorm2d-9           [-1, 16, 32, 32]              32
           Conv2d-10           [-1, 16, 32, 32]           2,304
      BatchNorm2d-11           [-1, 16, 32, 32]              32
           Fx2_16-12           [-1, 16, 32, 32]               0
           Conv2d-13           [-1, 16, 32, 32]           2,304
      BatchNorm2d-14           [-1, 16,

In [ ]:
MyModel()

MyModel(
  (conv1): Sequential(
    (0): Conv2d(3, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
    (1): BatchNorm2d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  )
  (conv2): conv16(
    (Fx1): Fx2_16(
      (conv1): Sequential(
        (0): Conv2d(16, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (1): BatchNorm2d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv2): Sequential(
        (0): Conv2d(16, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (1): BatchNorm2d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (residual): Sequential()
    )
    (Fx2): Fx2_16(
      (conv1): Sequential(
        (0): Conv2d(16, 16, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
        (1): BatchNorm2d(16, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      )
      (conv2): Sequential(
        (0)

# Train

In [9]:
# #get_model when loading model parameters trained w/ multi-gpu
def get_model(model_name, checkpoint_path):
    print("start loading model parameter")
    model = model_name()
    state_dict = torch.load(checkpoint_path)

    new_state_dict = OrderedDict()

    for k, v in state_dict.items():
        name = k.replace('module.', '')
        new_state_dict[name] = v
    model.load_state_dict(new_state_dict)
    print("model parameter loaded")
    return model

In [10]:

def train(n_epochs, train_loader) :
 
    epochs = n_epochs
    train_dataloader = train_loader
    #get current time -> folder name
    train_date = time.strftime("%y%m%d_%H%M", time.localtime(time.time()))
    # print(train_date)

    #log path for tensorboard
    log_path = os.path.join(os.getcwd(),"logs",train_date)
    print("log_path is {}".format(log_path))

    #model path to save '.pth'file
    model_path = os.path.join(os.getcwd(),"models",train_date)
    print("model_path is {}".format(model_path))
    if os.path.isdir(model_path) == False:
            os.makedirs(model_path)

    #writer for tensorboard
    writer = SummaryWriter(log_path)
 
    # whether available GPU
    # device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    #device = torch.device("mps") if torch.backends.mps.is_available() else "cpu"
    
    #define model (처음부터 학습시)
    print("처음부터 학습")
    model = MyModel()
    if torch.cuda.device_count() > 1:
        model = nn.DataParallel(model)
    model.to(device)

    ####### 이전 학습한거에 추가적으로 학습시
    # print("이전에 하던거 부터 학습")
    # model_name = MyModel
    # # /content/models/230412_2145/0.pth
    # checkpoint_path = './12.pth'  #수정 필요
    # model = get_model(model_name,checkpoint_path)
    # model.to(device)

    #######

    #define loss funcion, optimizer, (scheduler)
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.SGD(model.parameters(),lr=0.1, momentum=0.9,weight_decay=0.0001)
    #lr = 0.025, momentum = 0.9, weight_decay = 0.0005(original)
    # scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones = [81,122], gamma = 0.1)
    scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, milestones = [81,122], gamma = 0.1)
    # 0.1 -> 0.01 (32K iterations = 81 epochs) -> 0.001 (48K iterations = 122 epochs)

    print("start training")
    #Train
    for epoch in range(epochs):

        model.train(True)
        len_trainbatch = len(train_dataloader)

        i = 0
        for inputs, targets in tqdm(train_dataloader):
            i = i+1
            inputs, targets = inputs.to(device), targets.to(device)
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            

            writer.add_scalar('train/loss', loss.item(), epoch*len_trainbatch+i) #iterations
        
        scheduler.step()    
        print("train loss of {}th epoch is {}".format(epoch,loss.item())) ## 수정 필요
 
        # Save every N(30) epoch
        save_epoch = 30
        if save_epoch > 0 and epoch % save_epoch == save_epoch-1:
            state_dict = model.state_dict()
            torch.save(state_dict, os.path.join(model_path, str(epoch+1)+'.pth')) #epoch+1로 바꾸기
        
    print('Finished Training')

    #save model after training
    torch.save(model.state_dict(), './model.pth') 

   

# Main

In [11]:

def main() :

    random.seed(53)
    #train(163, trainloader) #64k iterations = 163 epochs
    train(163, trainloader) #64k iterations = 163 epochs

   

if __name__ == '__main__':
    main()

log_path is /content/logs/230417_0135
model_path is /content/models/230417_0135
처음부터 학습
start training


100%|██████████| 391/391 [00:43<00:00,  8.99it/s]


train loss of 0th epoch is 1.815954566001892


100%|██████████| 391/391 [00:37<00:00, 10.56it/s]


train loss of 1th epoch is 1.017509937286377


100%|██████████| 391/391 [00:38<00:00, 10.23it/s]


train loss of 2th epoch is 0.7631059288978577


100%|██████████| 391/391 [00:37<00:00, 10.42it/s]


train loss of 3th epoch is 0.7618404030799866


100%|██████████| 391/391 [00:37<00:00, 10.41it/s]


train loss of 4th epoch is 0.6331163644790649


100%|██████████| 391/391 [00:37<00:00, 10.44it/s]


train loss of 5th epoch is 0.9797455072402954


100%|██████████| 391/391 [00:37<00:00, 10.37it/s]


train loss of 6th epoch is 0.484571635723114


100%|██████████| 391/391 [00:37<00:00, 10.32it/s]


train loss of 7th epoch is 0.5337564945220947


100%|██████████| 391/391 [00:37<00:00, 10.38it/s]


train loss of 8th epoch is 0.6987408399581909


100%|██████████| 391/391 [00:37<00:00, 10.29it/s]


train loss of 9th epoch is 0.5212087631225586


100%|██████████| 391/391 [00:37<00:00, 10.38it/s]


train loss of 10th epoch is 0.4559572637081146


100%|██████████| 391/391 [00:37<00:00, 10.44it/s]


train loss of 11th epoch is 0.39366719126701355


100%|██████████| 391/391 [00:37<00:00, 10.52it/s]


train loss of 12th epoch is 0.3787572979927063


100%|██████████| 391/391 [00:37<00:00, 10.44it/s]


train loss of 13th epoch is 0.39672595262527466


100%|██████████| 391/391 [00:37<00:00, 10.36it/s]


train loss of 14th epoch is 0.4936068058013916


100%|██████████| 391/391 [00:37<00:00, 10.42it/s]


train loss of 15th epoch is 0.38852018117904663


100%|██████████| 391/391 [00:37<00:00, 10.43it/s]


train loss of 16th epoch is 0.1367660015821457


100%|██████████| 391/391 [00:37<00:00, 10.39it/s]


train loss of 17th epoch is 0.29003217816352844


100%|██████████| 391/391 [00:37<00:00, 10.35it/s]


train loss of 18th epoch is 0.4720730781555176


100%|██████████| 391/391 [00:37<00:00, 10.37it/s]


train loss of 19th epoch is 0.34230363368988037


100%|██████████| 391/391 [00:37<00:00, 10.43it/s]


train loss of 20th epoch is 0.28064820170402527


100%|██████████| 391/391 [00:37<00:00, 10.39it/s]


train loss of 21th epoch is 0.4017488360404968


100%|██████████| 391/391 [00:37<00:00, 10.42it/s]


train loss of 22th epoch is 0.21483704447746277


100%|██████████| 391/391 [00:37<00:00, 10.54it/s]


train loss of 23th epoch is 0.26761236786842346


100%|██████████| 391/391 [00:37<00:00, 10.51it/s]


train loss of 24th epoch is 0.35433366894721985


100%|██████████| 391/391 [00:37<00:00, 10.47it/s]


train loss of 25th epoch is 0.31735169887542725


100%|██████████| 391/391 [00:37<00:00, 10.39it/s]


train loss of 26th epoch is 0.1887972354888916


100%|██████████| 391/391 [00:37<00:00, 10.40it/s]


train loss of 27th epoch is 0.22285659611225128


100%|██████████| 391/391 [00:37<00:00, 10.41it/s]


train loss of 28th epoch is 0.4803212583065033


100%|██████████| 391/391 [00:37<00:00, 10.45it/s]


train loss of 29th epoch is 0.3925331234931946


100%|██████████| 391/391 [00:37<00:00, 10.35it/s]


train loss of 30th epoch is 0.3509537875652313


100%|██████████| 391/391 [00:37<00:00, 10.40it/s]


train loss of 31th epoch is 0.2767350375652313


100%|██████████| 391/391 [00:37<00:00, 10.37it/s]


train loss of 32th epoch is 0.16612648963928223


100%|██████████| 391/391 [00:37<00:00, 10.36it/s]


train loss of 33th epoch is 0.2277657687664032


100%|██████████| 391/391 [00:37<00:00, 10.43it/s]


train loss of 34th epoch is 0.17886677384376526


100%|██████████| 391/391 [00:37<00:00, 10.52it/s]


train loss of 35th epoch is 0.22561606764793396


100%|██████████| 391/391 [00:37<00:00, 10.55it/s]


train loss of 36th epoch is 0.3318116366863251


100%|██████████| 391/391 [00:37<00:00, 10.43it/s]


train loss of 37th epoch is 0.11670328676700592


100%|██████████| 391/391 [00:37<00:00, 10.42it/s]


train loss of 38th epoch is 0.27812284231185913


100%|██████████| 391/391 [00:37<00:00, 10.42it/s]


train loss of 39th epoch is 0.22766271233558655


100%|██████████| 391/391 [00:37<00:00, 10.38it/s]


train loss of 40th epoch is 0.4113788604736328


100%|██████████| 391/391 [00:37<00:00, 10.43it/s]


train loss of 41th epoch is 0.3038092255592346


100%|██████████| 391/391 [00:37<00:00, 10.39it/s]


train loss of 42th epoch is 0.15935243666172028


100%|██████████| 391/391 [00:37<00:00, 10.46it/s]


train loss of 43th epoch is 0.15894640982151031


100%|██████████| 391/391 [00:37<00:00, 10.44it/s]


train loss of 44th epoch is 0.14984987676143646


100%|██████████| 391/391 [00:37<00:00, 10.40it/s]


train loss of 45th epoch is 0.2516499161720276


100%|██████████| 391/391 [00:37<00:00, 10.33it/s]


train loss of 46th epoch is 0.26499244570732117


100%|██████████| 391/391 [00:37<00:00, 10.46it/s]


train loss of 47th epoch is 0.1023833155632019


100%|██████████| 391/391 [00:37<00:00, 10.38it/s]


train loss of 48th epoch is 0.29948097467422485


100%|██████████| 391/391 [00:37<00:00, 10.32it/s]


train loss of 49th epoch is 0.11893200874328613


100%|██████████| 391/391 [00:38<00:00, 10.28it/s]


train loss of 50th epoch is 0.15722398459911346


100%|██████████| 391/391 [00:37<00:00, 10.30it/s]


train loss of 51th epoch is 0.22448638081550598


100%|██████████| 391/391 [00:38<00:00, 10.26it/s]


train loss of 52th epoch is 0.10148818790912628


100%|██████████| 391/391 [00:38<00:00, 10.24it/s]


train loss of 53th epoch is 0.2909347414970398


100%|██████████| 391/391 [00:37<00:00, 10.33it/s]


train loss of 54th epoch is 0.29380935430526733


100%|██████████| 391/391 [00:38<00:00, 10.24it/s]


train loss of 55th epoch is 0.3131316006183624


100%|██████████| 391/391 [00:38<00:00, 10.21it/s]


train loss of 56th epoch is 0.2111687958240509


100%|██████████| 391/391 [00:37<00:00, 10.33it/s]


train loss of 57th epoch is 0.21338561177253723


100%|██████████| 391/391 [00:37<00:00, 10.37it/s]


train loss of 58th epoch is 0.1768362820148468


100%|██████████| 391/391 [00:37<00:00, 10.43it/s]


train loss of 59th epoch is 0.18624067306518555


100%|██████████| 391/391 [00:37<00:00, 10.35it/s]


train loss of 60th epoch is 0.23087139427661896


100%|██████████| 391/391 [00:37<00:00, 10.32it/s]


train loss of 61th epoch is 0.17866620421409607


100%|██████████| 391/391 [00:37<00:00, 10.40it/s]


train loss of 62th epoch is 0.2379215508699417


100%|██████████| 391/391 [00:37<00:00, 10.42it/s]


train loss of 63th epoch is 0.2086414098739624


100%|██████████| 391/391 [00:37<00:00, 10.42it/s]


train loss of 64th epoch is 0.24513384699821472


100%|██████████| 391/391 [00:37<00:00, 10.39it/s]


train loss of 65th epoch is 0.10319799184799194


100%|██████████| 391/391 [00:37<00:00, 10.43it/s]


train loss of 66th epoch is 0.12201546132564545


100%|██████████| 391/391 [00:37<00:00, 10.37it/s]


train loss of 67th epoch is 0.15774424374103546


100%|██████████| 391/391 [00:37<00:00, 10.44it/s]


train loss of 68th epoch is 0.11219145357608795


100%|██████████| 391/391 [00:37<00:00, 10.37it/s]


train loss of 69th epoch is 0.1737499237060547


100%|██████████| 391/391 [00:37<00:00, 10.37it/s]


train loss of 70th epoch is 0.15060319006443024


100%|██████████| 391/391 [00:37<00:00, 10.32it/s]


train loss of 71th epoch is 0.24748198688030243


100%|██████████| 391/391 [00:37<00:00, 10.38it/s]


train loss of 72th epoch is 0.35894954204559326


100%|██████████| 391/391 [00:37<00:00, 10.38it/s]


train loss of 73th epoch is 0.1158488541841507


100%|██████████| 391/391 [00:37<00:00, 10.39it/s]


train loss of 74th epoch is 0.22376656532287598


100%|██████████| 391/391 [00:37<00:00, 10.35it/s]


train loss of 75th epoch is 0.16685064136981964


100%|██████████| 391/391 [00:37<00:00, 10.41it/s]


train loss of 76th epoch is 0.17620885372161865


100%|██████████| 391/391 [00:37<00:00, 10.37it/s]


train loss of 77th epoch is 0.20834282040596008


100%|██████████| 391/391 [00:37<00:00, 10.32it/s]


train loss of 78th epoch is 0.2924129366874695


100%|██████████| 391/391 [00:37<00:00, 10.37it/s]


train loss of 79th epoch is 0.2532643973827362


100%|██████████| 391/391 [00:37<00:00, 10.34it/s]


train loss of 80th epoch is 0.11793837696313858


100%|██████████| 391/391 [00:38<00:00, 10.28it/s]


train loss of 81th epoch is 0.09151400625705719


100%|██████████| 391/391 [00:38<00:00, 10.27it/s]


train loss of 82th epoch is 0.09707753360271454


100%|██████████| 391/391 [00:37<00:00, 10.40it/s]


train loss of 83th epoch is 0.09156068414449692


100%|██████████| 391/391 [00:37<00:00, 10.41it/s]


train loss of 84th epoch is 0.10150317847728729


100%|██████████| 391/391 [00:37<00:00, 10.38it/s]


train loss of 85th epoch is 0.05088431388139725


100%|██████████| 391/391 [00:37<00:00, 10.37it/s]


train loss of 86th epoch is 0.018042175099253654


100%|██████████| 391/391 [00:37<00:00, 10.31it/s]


train loss of 87th epoch is 0.03758716210722923


100%|██████████| 391/391 [00:37<00:00, 10.46it/s]


train loss of 88th epoch is 0.011652404442429543


100%|██████████| 391/391 [00:37<00:00, 10.49it/s]


train loss of 89th epoch is 0.01896553859114647


100%|██████████| 391/391 [00:37<00:00, 10.56it/s]


train loss of 90th epoch is 0.09670460969209671


100%|██████████| 391/391 [00:36<00:00, 10.58it/s]


train loss of 91th epoch is 0.0068093566223979


100%|██████████| 391/391 [00:37<00:00, 10.55it/s]


train loss of 92th epoch is 0.029759928584098816


100%|██████████| 391/391 [00:37<00:00, 10.51it/s]


train loss of 93th epoch is 0.019955897703766823


100%|██████████| 391/391 [00:37<00:00, 10.41it/s]


train loss of 94th epoch is 0.029485564678907394


100%|██████████| 391/391 [00:37<00:00, 10.51it/s]


train loss of 95th epoch is 0.01960131525993347


100%|██████████| 391/391 [00:37<00:00, 10.54it/s]


train loss of 96th epoch is 0.007964463904500008


100%|██████████| 391/391 [00:37<00:00, 10.45it/s]


train loss of 97th epoch is 0.008513856679201126


100%|██████████| 391/391 [00:37<00:00, 10.51it/s]


train loss of 98th epoch is 0.03720250353217125


100%|██████████| 391/391 [00:37<00:00, 10.47it/s]


train loss of 99th epoch is 0.013946928083896637


100%|██████████| 391/391 [00:37<00:00, 10.50it/s]


train loss of 100th epoch is 0.0019423844059929252


100%|██████████| 391/391 [00:37<00:00, 10.51it/s]


train loss of 101th epoch is 0.009851117618381977


100%|██████████| 391/391 [00:37<00:00, 10.49it/s]


train loss of 102th epoch is 0.0014479558449238539


100%|██████████| 391/391 [00:37<00:00, 10.52it/s]


train loss of 103th epoch is 0.013956956565380096


100%|██████████| 391/391 [00:37<00:00, 10.48it/s]


train loss of 104th epoch is 0.017311256378889084


100%|██████████| 391/391 [00:37<00:00, 10.48it/s]


train loss of 105th epoch is 0.0077486587688326836


100%|██████████| 391/391 [00:37<00:00, 10.47it/s]


train loss of 106th epoch is 0.012543636374175549


100%|██████████| 391/391 [00:37<00:00, 10.47it/s]


train loss of 107th epoch is 0.006964600179344416


100%|██████████| 391/391 [00:37<00:00, 10.49it/s]


train loss of 108th epoch is 0.002007268136367202


100%|██████████| 391/391 [00:37<00:00, 10.50it/s]


train loss of 109th epoch is 0.006129803601652384


100%|██████████| 391/391 [00:37<00:00, 10.36it/s]


train loss of 110th epoch is 0.0028228447772562504


100%|██████████| 391/391 [00:37<00:00, 10.47it/s]


train loss of 111th epoch is 0.010988658294081688


100%|██████████| 391/391 [00:37<00:00, 10.45it/s]


train loss of 112th epoch is 0.0015440593706443906


100%|██████████| 391/391 [00:37<00:00, 10.45it/s]


train loss of 113th epoch is 0.008341727778315544


100%|██████████| 391/391 [00:37<00:00, 10.45it/s]


train loss of 114th epoch is 0.00240239966660738


100%|██████████| 391/391 [00:37<00:00, 10.49it/s]


train loss of 115th epoch is 0.015388241037726402


100%|██████████| 391/391 [00:37<00:00, 10.52it/s]


train loss of 116th epoch is 0.00797326397150755


100%|██████████| 391/391 [00:37<00:00, 10.50it/s]


train loss of 117th epoch is 0.010511280968785286


100%|██████████| 391/391 [00:37<00:00, 10.46it/s]


train loss of 118th epoch is 0.004131005145609379


100%|██████████| 391/391 [00:37<00:00, 10.48it/s]


train loss of 119th epoch is 0.004112566821277142


100%|██████████| 391/391 [00:37<00:00, 10.51it/s]


train loss of 120th epoch is 0.0038607928436249495


100%|██████████| 391/391 [00:37<00:00, 10.45it/s]


train loss of 121th epoch is 0.004147501662373543


100%|██████████| 391/391 [00:37<00:00, 10.50it/s]


train loss of 122th epoch is 0.0022129316348582506


100%|██████████| 391/391 [00:37<00:00, 10.45it/s]


train loss of 123th epoch is 0.024692459031939507


100%|██████████| 391/391 [00:37<00:00, 10.55it/s]


train loss of 124th epoch is 0.007152312435209751


100%|██████████| 391/391 [00:37<00:00, 10.53it/s]


train loss of 125th epoch is 0.0021845167502760887


100%|██████████| 391/391 [00:37<00:00, 10.39it/s]


train loss of 126th epoch is 0.020466729998588562


100%|██████████| 391/391 [00:37<00:00, 10.49it/s]


train loss of 127th epoch is 0.0033533982932567596


100%|██████████| 391/391 [00:37<00:00, 10.49it/s]


train loss of 128th epoch is 0.0034712410997599363


100%|██████████| 391/391 [00:37<00:00, 10.47it/s]


train loss of 129th epoch is 0.0030003334395587444


100%|██████████| 391/391 [00:37<00:00, 10.50it/s]


train loss of 130th epoch is 0.0037176422774791718


100%|██████████| 391/391 [00:37<00:00, 10.50it/s]


train loss of 131th epoch is 0.0050097620114684105


100%|██████████| 391/391 [00:37<00:00, 10.53it/s]


train loss of 132th epoch is 0.000510759127791971


100%|██████████| 391/391 [00:37<00:00, 10.48it/s]


train loss of 133th epoch is 0.0004633396747522056


100%|██████████| 391/391 [00:37<00:00, 10.46it/s]


train loss of 134th epoch is 0.002944567706435919


100%|██████████| 391/391 [00:37<00:00, 10.47it/s]


train loss of 135th epoch is 0.0344269722700119


100%|██████████| 391/391 [00:37<00:00, 10.48it/s]


train loss of 136th epoch is 0.01120397262275219


100%|██████████| 391/391 [00:37<00:00, 10.50it/s]


train loss of 137th epoch is 0.019425470381975174


100%|██████████| 391/391 [00:37<00:00, 10.49it/s]


train loss of 138th epoch is 0.0034824747126549482


100%|██████████| 391/391 [00:37<00:00, 10.46it/s]


train loss of 139th epoch is 0.0037510269321501255


100%|██████████| 391/391 [00:37<00:00, 10.46it/s]


train loss of 140th epoch is 0.0017724450444802642


100%|██████████| 391/391 [00:37<00:00, 10.46it/s]


train loss of 141th epoch is 0.0008882078109309077


100%|██████████| 391/391 [00:37<00:00, 10.41it/s]


train loss of 142th epoch is 0.0021282273810356855


100%|██████████| 391/391 [00:37<00:00, 10.45it/s]


train loss of 143th epoch is 0.010046787559986115


100%|██████████| 391/391 [00:37<00:00, 10.48it/s]


train loss of 144th epoch is 0.018355675041675568


100%|██████████| 391/391 [00:37<00:00, 10.46it/s]


train loss of 145th epoch is 0.015254025347530842


100%|██████████| 391/391 [00:37<00:00, 10.46it/s]


train loss of 146th epoch is 0.0013920030323788524


100%|██████████| 391/391 [00:37<00:00, 10.44it/s]


train loss of 147th epoch is 0.01651998609304428


100%|██████████| 391/391 [00:37<00:00, 10.49it/s]


train loss of 148th epoch is 0.0106661356985569


100%|██████████| 391/391 [00:37<00:00, 10.42it/s]


train loss of 149th epoch is 0.001348360674455762


100%|██████████| 391/391 [00:37<00:00, 10.49it/s]


train loss of 150th epoch is 0.004559732507914305


100%|██████████| 391/391 [00:37<00:00, 10.46it/s]


train loss of 151th epoch is 0.004803747870028019


100%|██████████| 391/391 [00:37<00:00, 10.47it/s]


train loss of 152th epoch is 0.017624296247959137


100%|██████████| 391/391 [00:37<00:00, 10.45it/s]


train loss of 153th epoch is 0.0023684801999479532


100%|██████████| 391/391 [00:37<00:00, 10.51it/s]


train loss of 154th epoch is 0.002646071370691061


100%|██████████| 391/391 [00:37<00:00, 10.49it/s]


train loss of 155th epoch is 0.005809304304420948


100%|██████████| 391/391 [00:37<00:00, 10.48it/s]


train loss of 156th epoch is 0.0010645684087648988


100%|██████████| 391/391 [00:37<00:00, 10.45it/s]


train loss of 157th epoch is 0.010038428008556366


100%|██████████| 391/391 [00:37<00:00, 10.36it/s]


train loss of 158th epoch is 0.0040747602470219135


100%|██████████| 391/391 [00:37<00:00, 10.39it/s]


train loss of 159th epoch is 0.0031118388287723064


100%|██████████| 391/391 [00:37<00:00, 10.46it/s]


train loss of 160th epoch is 0.0062148841097950935


100%|██████████| 391/391 [00:37<00:00, 10.46it/s]


train loss of 161th epoch is 0.002089728834107518


100%|██████████| 391/391 [00:37<00:00, 10.45it/s]

train loss of 162th epoch is 0.002325118985027075
Finished Training


# Test

- model

100 100 Loss: 0.306 | Acc: 93.030% (9303/10000)

- 30

100 100 Loss: 0.436 | Acc: 85.740% (8574/10000)

- 60 

100 100 Loss: 0.438 | Acc: 86.810% (8681/10000)

- 90

100 100 Loss: 0.259 | Acc: 92.730% (9273/10000)

- 120

100 100 Loss: 0.306 | Acc: 92.720% (9272/10000)

- 150

100 100 Loss: 0.307 | Acc: 92.970% (9297/10000)

In [24]:
def test():

    ####### not touch
    model_name = MyModel
    # /content/models/230412_2145/0.pth
    #/content/models/230412_2208/1.pth
    #/content/models/230416_0841/90.pth
    #/content/models/230417_0135/30.pth
    checkpoint_path = './models/230417_0135/150.pth'  #수정 필요
    mode = 'test' 
#     data_dir = "./test_data"
#     meta_path = "./answer.json"
    model = get_model(model_name,checkpoint_path)
    #######
    
#     # Create training and validation datasets
#     test_datasets = MyDataset(meta_path, data_dir, data_transforms['test'],is_train= False) #self,meta_path,root_dir,transform=None, is_train = True

#     # Create training and validation dataloaders
#     test_dataloader = torch.utils.data.DataLoader(test_datasets, batch_size=batch_size, shuffle=False, num_workers=4)

    test_set = torchvision.datasets.CIFAR10(
    root='./data', train=False, download=True, transform=data_transforms['test'])
    
    testloader = torch.utils.data.DataLoader(
    test_set, batch_size=100, shuffle=False, num_workers=2)
    
    # Detect if we have a GPU available
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    # device = torch.device("mps") if torch.backends.mps.is_available() else "cpu"
    
    # Send the model to GPU
    #multi-GPU
    if torch.cuda.device_count() > 1:
        model = nn.DataParallel(model)
    model = model.to(device)

    # Set model as evaluation mode
    for param in model.parameters():
        param.requires_grad = False
    model.eval()
    
    # Inference
    test_loss = 0
    correct = 0
    total = 0
    criterion = nn.CrossEntropyLoss()
    with torch.no_grad():
        batch_idx = 0
        for inputs, targets in tqdm(testloader):
            inputs, targets = inputs.to(device), targets.to(device)
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            
            test_loss += loss.item()
            _, predicted = outputs.max(1)
            total += targets.size(0)
            correct += predicted.eq(targets).sum().item()
            batch_idx = batch_idx + 1
            print(batch_idx, len(testloader), 'Loss: %.3f | Acc: %.3f%% (%d/%d)'
                         % (test_loss/(batch_idx+1), 100.*correct/total, correct, total))



In [25]:

def main() :

    random.seed(53)
    test()

   
if __name__ == '__main__':
    main()

start loading model parameter
model parameter loaded
Files already downloaded and verified


  3%|▎         | 3/100 [00:00<00:12,  8.01it/s]

1 100 Loss: 0.123 | Acc: 92.000% (92/100)
2 100 Loss: 0.242 | Acc: 91.000% (182/200)
3 100 Loss: 0.229 | Acc: 91.667% (275/300)
4 100 Loss: 0.236 | Acc: 92.500% (370/400)


  8%|▊         | 8/100 [00:00<00:06, 14.54it/s]

5 100 Loss: 0.240 | Acc: 92.800% (464/500)
6 100 Loss: 0.228 | Acc: 93.333% (560/600)
7 100 Loss: 0.255 | Acc: 93.571% (655/700)
8 100 Loss: 0.301 | Acc: 92.750% (742/800)


 12%|█▏        | 12/100 [00:00<00:05, 16.60it/s]

9 100 Loss: 0.309 | Acc: 92.889% (836/900)
10 100 Loss: 0.312 | Acc: 92.700% (927/1000)
11 100 Loss: 0.316 | Acc: 92.455% (1017/1100)
12 100 Loss: 0.320 | Acc: 92.250% (1107/1200)


 16%|█▌        | 16/100 [00:01<00:05, 16.62it/s]

13 100 Loss: 0.307 | Acc: 92.308% (1200/1300)
14 100 Loss: 0.307 | Acc: 92.000% (1288/1400)
15 100 Loss: 0.299 | Acc: 92.200% (1383/1500)
16 100 Loss: 0.298 | Acc: 92.250% (1476/1600)


 20%|██        | 20/100 [00:01<00:04, 17.32it/s]

17 100 Loss: 0.297 | Acc: 92.412% (1571/1700)
18 100 Loss: 0.302 | Acc: 92.333% (1662/1800)
19 100 Loss: 0.314 | Acc: 92.316% (1754/1900)
20 100 Loss: 0.327 | Acc: 92.200% (1844/2000)


 25%|██▌       | 25/100 [00:01<00:04, 16.52it/s]

21 100 Loss: 0.325 | Acc: 92.190% (1936/2100)
22 100 Loss: 0.322 | Acc: 92.318% (2031/2200)
23 100 Loss: 0.331 | Acc: 92.043% (2117/2300)
24 100 Loss: 0.327 | Acc: 92.083% (2210/2400)
25 100 Loss: 0.326 | Acc: 92.200% (2305/2500)


 30%|███       | 30/100 [00:01<00:03, 18.13it/s]

26 100 Loss: 0.337 | Acc: 92.115% (2395/2600)
27 100 Loss: 0.331 | Acc: 92.222% (2490/2700)
28 100 Loss: 0.326 | Acc: 92.393% (2587/2800)
29 100 Loss: 0.327 | Acc: 92.379% (2679/2900)
30 100 Loss: 0.323 | Acc: 92.467% (2774/3000)


 33%|███▎      | 33/100 [00:02<00:03, 20.52it/s]

31 100 Loss: 0.323 | Acc: 92.516% (2868/3100)
32 100 Loss: 0.319 | Acc: 92.625% (2964/3200)
33 100 Loss: 0.315 | Acc: 92.667% (3058/3300)
34 100 Loss: 0.312 | Acc: 92.735% (3153/3400)
35 100 Loss: 0.316 | Acc: 92.600% (3241/3500)
36 100 Loss: 0.315 | Acc: 92.639% (3335/3600)


 40%|████      | 40/100 [00:02<00:02, 23.92it/s]

37 100 Loss: 0.319 | Acc: 92.541% (3424/3700)
38 100 Loss: 0.326 | Acc: 92.500% (3515/3800)
39 100 Loss: 0.323 | Acc: 92.590% (3611/3900)
40 100 Loss: 0.322 | Acc: 92.625% (3705/4000)
41 100 Loss: 0.327 | Acc: 92.488% (3792/4100)
42

 46%|████▌     | 46/100 [00:02<00:02, 24.59it/s]

 100 Loss: 0.326 | Acc: 92.524% (3886/4200)
43 100 Loss: 0.325 | Acc: 92.558% (3980/4300)
44 100 Loss: 0.326 | Acc: 92.568% (4073/4400)
45 100 Loss: 0.324 | Acc: 92.644% (4169/4500)
46 100 Loss: 0.324 | Acc: 92.652% (4262/4600)
47 100 Loss: 0.323 | Acc: 92.702% (4357/4700)


 52%|█████▏    | 52/100 [00:02<00:01, 26.15it/s]

48 100 Loss: 0.328 | Acc: 92.604% (4445/4800)
49 100 Loss: 0.322 | Acc: 92.694% (4542/4900)
50 100 Loss: 0.325 | Acc: 92.660% (4633/5000)
51 100 Loss: 0.322 | Acc: 92.706% (4728/5100)
52 100 Loss: 0.319 | Acc: 92.712% (4821/5200)
53 100 Loss: 0.320 | Acc: 92.660% (4911/5300)


 59%|█████▉    | 59/100 [00:02<00:01, 28.39it/s]

54 100 Loss: 0.318 | Acc: 92.685% (5005/5400)
55 100 Loss: 0.318 | Acc: 92.673% (5097/5500)
56 100 Loss: 0.317 | Acc: 92.714% (5192/5600)
57 100 Loss: 0.317 | Acc: 92.684% (5283/5700)
58 100 Loss: 0.315 | Acc: 92.724% (5378/5800)
59 100 Loss: 0.316 | Acc: 92.712% (5470/5900)


 65%|██████▌   | 65/100 [00:03<00:01, 26.77it/s]

60 100 Loss: 0.313 | Acc: 92.717% (5563/6000)
61 100 Loss: 0.311 | Acc: 92.770% (5659/6100)
62 100 Loss: 0.310 | Acc: 92.742% (5750/6200)
63 100 Loss: 0.308 | Acc: 92.794% (5846/6300)
64 100 Loss: 0.306 | Acc: 92.859% (5943/6400)
65 100 Loss: 0.307 | Acc: 92.831% (6034/6500)


 68%|██████▊   | 68/100 [00:03<00:01, 27.18it/s]

66 100 Loss: 0.306 | Acc: 92.818% (6126/6600)
67 100 Loss: 0.305 | Acc: 92.836% (6220/6700)
68 100 Loss: 0.306 | Acc: 92.824% (6312/6800)
69 100 Loss: 0.306 | Acc: 92.826% (6405/6900)
70 100 Loss: 0.305 | Acc: 92.843% (6499/7000)
71 100 Loss: 0.310 | Acc: 92.803% (6589/7100)
72 

 75%|███████▌  | 75/100 [00:03<00:00, 26.92it/s]

100 Loss: 0.311 | Acc: 92.792% (6681/7200)
73 100 Loss: 0.309 | Acc: 92.836% (6777/7300)
74 100 Loss: 0.310 | Acc: 92.851% (6871/7400)
75 100 Loss: 0.311 | Acc: 92.853% (6964/7500)
76 100 Loss: 0.310 | Acc: 92.882% (7059/7600)
77 100 Loss: 0.314 | Acc: 92.831% (7148/7700)


 82%|████████▏ | 82/100 [00:03<00:00, 26.66it/s]

78 100 Loss: 0.311 | Acc: 92.885% (7245/7800)
79 100 Loss: 0.310 | Acc: 92.911% (7340/7900)
80 100 Loss: 0.308 | Acc: 92.938% (7435/8000)
81 100 Loss: 0.307 | Acc: 92.963% (7530/8100)
82 100 Loss: 0.308 | Acc: 92.939% (7621/8200)


 89%|████████▉ | 89/100 [00:04<00:00, 27.74it/s]

83 100 Loss: 0.307 | Acc: 92.964% (7716/8300)
84 100 Loss: 0.306 | Acc: 92.976% (7810/8400)
85 100 Loss: 0.307 | Acc: 92.941% (7900/8500)
86 100 Loss: 0.310 | Acc: 92.884% (7988/8600)
87 100 Loss: 0.308 | Acc: 92.908% (8083/8700)
88 100 Loss: 0.309 | Acc: 92.886% (8174/8800)
89 100 Loss: 0.307 | Acc: 92.888% (8267/8900)


 95%|█████████▌| 95/100 [00:04<00:00, 26.34it/s]

90 100 Loss: 0.307 | Acc: 92.911% (8362/9000)
91 100 Loss: 0.307 | Acc: 92.912% (8455/9100)
92 100 Loss: 0.307 | Acc: 92.924% (8549/9200)
93 100 Loss: 0.310 | Acc: 92.882% (8638/9300)
94 100 Loss: 0.311 | Acc: 92.904% (8733/9400)
95 100 Loss: 0.310 | Acc: 92.916% (8827/9500)


100%|██████████| 100/100 [00:04<00:00, 22.06it/s]

96 100 Loss: 0.308 | Acc: 92.948% (8923/9600)
97 100 Loss: 0.306 | Acc: 92.979% (9019/9700)
98 100 Loss: 0.306 | Acc: 92.969% (9111/9800)
99 100 Loss: 0.307 | Acc: 92.970% (9204/9900)
100 100 Loss: 0.307 | Acc: 92.970% (9297/10000)
